### DISCUSSION TITLE
RFC: Include custom url matcher in filesystem routing

### DISCUSSION THREAD

**Discussion by DevWedeloper** [analogjs/analog#1452] [Ideas]
### Which scope/s are relevant/related to the feature request?

router

### Information

This RFC proposes adding support for **Custom URL Matchers** within the existing filesystem-based routing of Analog. This feature would allow developers to define custom logic for URL pattern matching directly within `.page.ts` files, enabling more flexible and complex routing patterns.

Proposed Implementation:

1. Naming Pattern for Custom Matchers: 
- Introduce a naming convention or symbol (e.g., `$`) in `.page.ts` filenames to signify that the file contains a custom URL matcher.
- For example, a file named `[$example].page.ts` would indicate that this route uses a custom matcher instead of relying on standard static or dynamic routing.

2. Matcher Defined Inside `.page.ts`:
- Developers would define a custom matcher function, similar to Angular’s UrlMatcher. This function would return a UrlMatchResult if the URL matches the defined logic.

   Example:
  ```
  export const urlMatcher = (segments: UrlSegment[], group: UrlSegmentGroup, route: Route): UrlMatchResult | null => {
     ...
  };
  
  ```

### Describe any alternatives/workarounds you're currently using

I'm using the provideRouter along with provideFileRouter

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

**Comment by brandonroberts** [comment]
I think manually defining an additional route that uses a custom matcher would work better for this use case. Using regex for a filesystem-based route seems like it would get confusing fast. We added support for providing extra routes here

https://github.com/analogjs/analog/pull/1446

**Reply by GuillaumeNury** [reply]
Hi @brandonroberts, I'm struggling to implement i18n-style routing.

I have a file-system routing setup that looks like this:

```md
/pages
  /(layout)
    /clients
      /index.page.ts
      /[name].page.ts
```

Depending on the language, I'd like to be able to access this route using `/clients` when language is EN or FR. However when the language is ES:
- `/clients` should return a 404
- `/clientes` should match the file-routed component

I have a workaround, with "extra routes" but I am not really happy with it. I have a route `/clientes` route that silently redirects to `/clients` without modifying the URL and I manually send a 404 when the user tries to access to the "hidden" `clients` route when the language is ES..

It would be awesome to do something like this, while still benefiting from layouts, server-side data loading, and everything else Analog has to offer:

```ts
export const routeMeta: RouteMeta = {
  matcher: i18nRouteMatcher({
    en: '/clients',
    fr: '/clients',
    es: '/clientes',
    de: '/kunden',
  })
};
```
